### ISSUE TITLE
middleware/cache serve stale records on timeout

### ISSUE THREAD

**Issue by stp-ip** [coredns/coredns#364]
Cloudflare seems to be moving that direction due to the dyn outage.  
With a LRU cache and enough memory one could hold most of the records in the cache and serve stale answers, when a refresh times out or failes.  
Hold records in cache as long as memory and LRU rules allow and refresh with backend on each query, if TTL is timed out.

**Comment by miekg** [comment]
Yes, but sometimes the TTL timeout is what is expected; DNSSEC keyrollovers; migrating A records etc. 

So you have to more careful, i.e. only allow timeout records to be returned when the upstream nameserver experience network timeouts.

Also in the case of CoreDNS' cache we have a number of different upstream, i.e. DNS ones, or etcd or kubernetes; do we want to implement the same behavior there? That would counter the shortlived TTLs of those environments.

As CoreDNS isn't a full DNS recursion, I'm not sure if this functionality belong here - but it is good to have a discussion and see what is possible - maybe even a separate middleware that handles it for.

**Comment by stp-ip** [comment]
I agree we shouldn't return a record with a timed out TTL. We should always check the backend first and if that check fails serve the stale record only.

The issue with using network timeouts is, that one tries to be smarter then the local resolver of the users. They might get a stale answer even so they might have successfully retried with another ns. So we might introduce issue during network partitions.

From a users pov I think, that the cache middleware should be the last to run and basically just cache everything, when enabled. Shouldn't matter, if it's etcd, k8s or file backend. (file backend should not need the cache, but it simplifies usage and implementation in my view)
So LRU cache, which caches every record, if enabled. Cache holds records as long as memory allows, but forces a recheck with the backend, if the TTL is timed out. Does that sound reasonable?

I wouldn't push serving stale records as a separate middleware. Just a configuration option in cache. Seems to connected to split that up.
CoreDNS doesn't need to have full recursion right? That's where the proxy middleware comes in. Maybe someone builds a recursion middleware so, but then cache could still just sit at the end and serve it.

**Comment by miekg** [comment]
the LRU cache you describe is how the cache currently works in CoreDNS-002 (modulo bugs).

> I agree we shouldn't return a record with a timed out TTL. We should always check the backend first and if that check fails serve the stale record only.
> 
> The issue with using network timeouts is, that one tries to be smarter then the local resolver of the users. They might get a stale answer even so they might have successfully retried with another ns. So we might introduce issue during network partitions.

But how is the cache going to know the backend timed out?

Related: DNSSEC signature lifetimes might throw a wrench in this as well

**Comment by stp-ip** [comment]
Took another look at the current cache middleware.
So currently the cache defines its own TTL, which might be fine, but can overwrite the TTL of the actual record. Basically we might already serve stale records.
The cache TTL could be set to a week and the records wouldn't be rechecked.

So maybe the suggestion is clearer saying that we should add rechecking of cached values into the cache. Basically the cache looks at the time the cache item was introduced and checks that against the TTL of the record. If the time it was added to the cache is older than the TTL of the record, the cache rechecks (tries to pull in the record - same path as for non cached records), if an answer is provided it will overwrite the current cache item. If there is an error or timeout, the cache middleware could check periodically (maybe with a backing off timer) for changes to the backend availability. It would still serve the stale answers.

Current:
- cache TTL is used to invalidate a cache item.

Suggestion:
- cache TTL can overwrite record TTL.
- record TTL is used for cache invalidation.
- on invalidation requests will initiate a new pull, but serve stale records on backend error.
- define a max_stale timeout to have a specific cut-off for old records (may default to a few days).
- invalidated items are removed from cache, when max_stale timeout is reached or the backend answers with NXDOMAIN or similar.

Haven't looked at the signature lifetimes, but I assume that could be another invalidation step to check for?
Basically checking the signature lifetime and if it has run out, remove the item from the cache immediately (so not just an invalidation) and go the usual pull from backend route as would be necessary. (Does that already happen with a cache TTL set to a longer time?)

Thanks for bearing with my suggestions and questions as always.

**Comment by miekg** [comment]
[ Quoting notifications@github.com in "Re: [miekg/coredns] middleware/cach..." ]

> So maybe the suggestion is clearer saying that we should add rechecking of cached values into the cache. Basically the cache looks at the time the cache item was introduced and checks that against the TTL of the record. If the time it was added to the cache is older than the TTL of the record, the cache rechecks (tries to pull in the record - same path as for non cached records), if an answer is provided it will overwrite the current cache item. If there is an error or timeout, the cache middleware could check periodically (maybe with a backing off timer) for changes to the backend availability. It would still serve the stale answers.

That should atually not be that hard now you write it like this.

> Current:
> - cache TTL is used to invalidate a cache item.
> 
> Suggestion:
> - cache TTL can overwrite record TTL.

If you set the TTL in cache stanza to 0 the "packets" TTL is used.

> - record TTL is used for cache invalidation.
> - on invalidation requests will initiate a new pull, but serve stale records on backend error.
> - define a max_stale timeout to have a specific cut-off for old records (may default to a few days).
> - invalidated items are removed from cache, when max_stale timeout is reached or the backend answers with NXDOMAIN or similar.

Ack. Care to implement something like this? :-)

> Haven't looked at the signature lifetimes, but I assume that could be another invalidation step to check for?
> Basically checking the signature lifetime and if it has run out, remove the item from the cache immediately (so not just an invalidation) and go the usual pull from backend route as would be necessary. (Does that already happen with a cache TTL set to a longer time?)

Current cache also does not care about sig lifetimes. We might get away with
just postponing this and assuming it will still validate.

> Thanks for bearing with my suggestions and questions as always.

Np. discussion like these are good to have.

**Comment by stp-ip** [comment]
> > Current:
> > - cache TTL is used to invalidate a cache item.
> >   Suggestion:
> > - cache TTL can overwrite record TTL.
> 
> If you set the TTL in cache stanza to 0 the "packets" TTL is used.

So if it is set to 0 or not set, it'll use the record TTL? (I assume that's how it should be, but in the docs it says, it uses 1h and 1/2h depending on the backend answer)

> > - record TTL is used for cache invalidation.
> > - on invalidation requests will initiate a new pull, but serve stale records on backend error.
> > - define a max_stale timeout to have a specific cut-off for old records (may default to a few days).
> > - invalidated items are removed from cache, when max_stale timeout is reached or the backend answers with NXDOMAIN or similar.
> 
> Ack. Care to implement something like this? :-)

I would love to, but I know the current time wouldn't allow me to commit to something like that. Therefore up for graps ;) 

> > Haven't looked at the signature lifetimes, but I assume that could be another invalidation step to check for?
> > Basically checking the signature lifetime and if it has run out, remove the item from the cache immediately (so not just an invalidation) and go the usual pull from backend route as would be necessary. (Does that already happen with a cache TTL set to a longer time?)
> 
> Current cache also does not care about sig lifetimes. We might get away with
> just postponing this and assuming it will still validate.

I'll open an issue to keep the issue documented. #367

**Comment by stp-ip** [comment]
Clarification for anyone coming later (might be added to the docs later on):
no cache -> records are not cached and record TTL is provided
`cache` -> records are cached using record TTL up to 3600 or 1800 depending on backend answer
`cache 0` -> records are cached with TTL 5
`cache 300` -> records are cached using record TTL as long as has a smaller value than 300 (cache TTL)

**Comment by miekg** [comment]
[ Quoting notifications@github.com in "Re: [miekg/coredns] middleware/cach..." ]

> Clarification for anyone coming later (might be added to the docs later on):
> no cache -> records are not cached and record TTL is provided
> `cache` -> records are cached using record TTL up to 3600 or 1800 depending on backend answer
> `cache 0` -> records are cached with TTL 5
> `cache 300` -> records are cached using record TTL as long as has a smaller value than 300 (cache TTL)

Would be good to open a PR for that and add it to the docs.

/Miek

## 

Miek Gieben

**Comment by stp-ip** [comment]
Will do as soon as the more pressing issues are fixed and the feature set seems to be stable.
